src/qmapshack/canvas/CCanvas.cpp

```diff
@@ -1133,19 +1133,19 @@ void CCanvas::displayInfo(const QPoint& px)
     QToolTip::hideText();
 }
 
-void CCanvas::findPoiCloseBy(const QPoint& px, QSet<IPoiItem>& poiItems, QList<QPointF>& posPoiHighlight)  const
+void CCanvas::findPoiCloseBy(const QPoint& px, QSet<const IPoiItem*>& poiItems, QList<QPointF>& posPoiHighlight)  const
```

Comment:
passing pointers a long methods as return values is super risky.  What happens to `poiItems`? Will it be stored? You can't tell. What happens to the objects pointed to in `poiItems`? You can't tell. They can be destroyed at any time and leaving a set of dangling pointers.

Reply:
My idea was that the `IPoiFile` object "owns" the `IPoiItem` it returns and will take care of its deletion when the file is closed and keeps it stored until then. The pointers should not be stored anywhere, although that is not bullet-proof and I have to check whether I've been true to that.

What are the other options though? const references can't be put into QSets afaik and handing down the items by value creates unnecessary copies all along the way. Shared Pointers would be an option, but as far as I remember you weren't too fond of these :smiley:

Reply:
The policy so far is (If I recall correctly, feel free to find examples contradicting):

* Pass references if possible.
* If references are not possible, pass copies.

We do have a very complex GUI in QMapShack, allowing the user to do a lot of stuff out of the order we developers expect. And there are a bunch of developers hacking the code. So we really should take care to make the code bulletproof. 

For example all GIS items are passed along by their keys, never by a pointer. If I want a pointer I have to use the key to get access to the object and I have to check if I really get an object. By that an item can be deleted while I still keep that list of keys.

If the additional copies become a problem we have to think about a better approach/concept. Passing pointers within a container is always like being blind folded and making a step towards the abyss. The first few steps might be ok but there will be the last fatal step sooner or later.

As it previously has been copies - and I think this is because of the old Garmin IMG based POI code uses copies - I would say that past-me has made the decision that the number of copies is countable and it's ok.

Reply:
> As it previously has been copies - and I think this is because of the old Garmin IMG based POI code uses copies - I would say that past-me has made the decision that the number of copies is countable and it's ok

I think there number of POIs possibly being on screen is much higher on outer zoom levels than those that come from a garmin map. Also, the items changed from being a relatively small struct to a full polymorphic class with a probably larger footprint, So I think we should go for the key based approach then, as the items have a key anyway.